_posts/2018-02-27-optimizing-grpc-part-1.md

```diff
@@ -0,0 +1,395 @@
+published: false
+permalink: blog/optimizing-grpc-part-1
+author: Carl Mastrangelo
+company: Google
+company-link: https://www.google.com
+---
+
+# Optimizing gRPC - Part 1
+
+A common question with gRPC is how to make it fast.  The gRPC library offers users access to high
+performance RPCs, but it isn't always clear how to achieve this.  Because this question is common 
+enough I thought I would try to show my thought process when tuning programs.
+
+## Setup
+
+Consider a basic Key Value service that is used by multiple other programs.  The service needs to
+be safe for concurrent access in case updates happen at the same time.  It needs to be able to scale
+up to use the available hardware.   Lastly, it needs to be fast.  gRPC is a perfect fit for this
+kind of service, so let's use it as our starting point.
+
+For the sake of example, I have written a
+[client and server](https://github.com/carl-mastrangelo/kvstore) using gRPC Java. The program is
+split into three main classes, and a protobuf file describing the API:
+
+* [KvClient](https://github.com/carl-mastrangelo/kvstore/blob/01-start/src/main/java/io/grpc/examples/KvClient.java)
+  is a simulated user of the key value system.   It randomly creates, retrieves, updates,
+  and deletes keys and values.  The size of keys and values it uses is also randomly decided
+  using an [exponential distribution](https://en.wikipedia.org/wiki/Exponential_distribution).
+* [KvService](https://github.com/carl-mastrangelo/kvstore/blob/01-start/src/main/java/io/grpc/examples/KvService.java)
+  is an implementation of the key value service.  It is installed by the gRPC Server to handle
+  the requests issued by the client.  To simulate storing the keys and values on disk, it adds
+  short sleeps while handling the request.  Reads and writes will experience a 10 and 50
+  millisecond delay to make the example act more like a persistent database.
+* [KvRunner](https://github.com/carl-mastrangelo/kvstore/blob/01-start/src/main/java/io/grpc/examples/KvRunner.java)
+  orchestrates the interaction between the client and the server.  It is the main entry point,
+  starting both the client and server in process, and waiting for the the client to execute its
+  work.  The runner does work for 60 seconds and then records how many RPCs were completed.
+* [kvstore.proto](https://github.com/carl-mastrangelo/kvstore/blob/01-start/src/main/proto/kvstore.proto)
+  is the protocol buffer definition of our service.  It describes exactly what clients can expect
+  from the service. For the sake of simplicity, we will use Create, Retrieve, Update, and Delete
+  as the operations (commonly known as CRUD).  These operations work with keys and values made up
+  of arbitrary bytes.  While they are somewhat REST like, we reserve the right to diverge and
+  add more complex operations in the future.
+
+While Protocol buffers (protos) aren't a requirement of using gRPC, they are convenient.  We will
```

Comment:
[protocol buffers](https://developers.google.com/protocol-buffers/)(protos) aren't required to use gRPC, they are a very convenient way to define service interfaces and generate client and server code. The generated code acts as glue code between the application logic and the core gRPC library. We call the code called by a gRPC client the _stub_.

Reply:
Done, but avoided using the word "call" twice in the last sentence.